### ISSUE TITLE
Memory leak in `ipfs add`

### ISSUE THREAD

**Issue by lgierth** [ipfs/go-ipfs#1924]
I'm add'ing https://downloads.openwrt.org/chaos_calmer/15.05/

```sh
lars@castor:~$ find downloads.openwrt.org/chaos_calmer/15.05/ | wc -l
278636
lars@castor:~$ du -sh downloads.openwrt.org/chaos_calmer/15.05/
37G	downloads.openwrt.org/chaos_calmer/15.05/
```

It ran for about 1h20m, in the meantime Go resident memory went from 121M to 13261M, and then it probably OOMed.

These are a lot of files that it needs to keep hashes for in memory, but it looks more like we're leaking the file contents.

I just started another run and so far it's consuming 3761M after adding 1870M worth of files. Here's the pprof goroutine, goroutine-full, and heap dumps: https://ipfs.io/ipfs/QmYCkvATcJcpZbs4mwvMZXzNp2CPKZKBkawYEBpZnv4Xbq

Aborting the second run now.

cc @whyrusleeping

**Comment by jbenet** [comment]
@whyrusleeping think this is the dag editor accumulating stuff in resident memory for a long time?

**Comment by whyrusleeping** [comment]
@lgierth could I also get your binary?

**Comment by lgierth** [comment]
Sure, the CLI is 0.3.8 from https://gobuilder.me/get/github.com/ipfs/go-ipfs/cmd/ipfs/ipfs_v0.3.8_linux-amd64.zip and the daemon is 2afe4a4d237234d27281eb5447c26119e5fe355a ([uploaded here](https://ipfs.io/ipfs/QmWiZeq8JHx7U21Szhh4cbJD1wrBStiFHv2uSCo1DZC4Ve))

**Comment by rht** [comment]
This was reported in #1584 (i.e. this issue's reference count is 2).

**Comment by rht** [comment]
@lgierth I think this has been fixed since #1965. There will be less memory growth once #1964 is merged.

**Comment by lgierth** [comment]
Still leaks, @davidar's add is eating 2.5G/h right now:

![graph](https://ipfs.io/ipfs/QmcoLN6Ve9YX3aUShdoQshqcEKbRofEeRyPe8aztzRHWTW/screencap.2016-05-30T00:56:39Z.png)

**Comment by lgierth** [comment]
Interestingly, the other metric being affected by this leak is the number of peers:

![peers graph](https://ipfs.io/ipfs/QmVM72usVabJRnfbFbmr5Hoyts4EYEdF3e7xh9LF8MfcaY/screencap.2016-05-30T01:26:41Z.png)

**Comment by lgierth** [comment]
Debugging info from #2615:

- pprof: https://ipfs.io/ipfs/QmedpSbkZxUQQpwRiM2eVdGWUCnDZsr51YtRLdv4UdFANv
- stack dump: https://ipfs.io/ipfs/QmfTmdf8r7g9X6BW3gEwfKrdvPdBFijdSiCDWAMew7aThz

**Comment by rht** [comment]
This is because the mfs abstraction still creates intermediate root dag for each `mfs.PutNode`.
The memory profile as of https://github.com/ipfs/go-ipfs/commit/6c3350369e82809c0fdcedd4d3d2e3f1d5d7111c:
![memory](https://cloud.githubusercontent.com/assets/395821/15641634/7ec14320-266c-11e6-8161-5e2106e1f1a6.png)

One fix is to repurpose https://github.com/ipfs/go-ipfs/pull/1964 into a new function in mfs that doesn't construct any intermediate root dag.
Simplifying the function to roughly just contain `r.dserv.Add` results in a flatter graph:
![memory](https://cloud.githubusercontent.com/assets/395821/15641798/6c800eca-266d-11e6-8852-76dafe645ae0.png)
(and a throughput rate that is slightly faster than git / tar|gz at files of >10kb range)

**Comment by whyrusleeping** [comment]
> This is because the mfs abstraction still creates intermediate root dag for each mfs.PutNode.

This is a regression, introduced in 06e71033. There is a fix for it in https://github.com/ipfs/go-ipfs/pull/2690 but it appears interest there has... languished.

**Comment by rht** [comment]
> This is a regression, introduced in https://github.com/ipfs/go-ipfs/commit/06e710330a1ce6e937fa214998d8da5cbf5a8555.
This assumes
1. there is no regression before the commit / pr.
2. `mfs.PutNode` is fixed by caching
Neither is the case. I reverted the https://github.com/ipfs/go-ipfs/commit/06e710330a1ce6e937fa214998d8da5cbf5a8555 and ran the memory profiling benchmark, still observed the leak.

**Comment by lgierth** [comment]
@davidar restarted his add command and biham is back to 29G, i dumped the debugging info into biham:/root/memleak.

**Comment by whyrusleeping** [comment]
This should be resolved by #2795. Its ready to go, just watiing for a small bit of CR first

**Comment by mattseh** [comment]
Hi,

I brought this up before in https://github.com/ipfs/go-ipfs/issues/2615 - This time I took the 64 bit Linux v0.4.3 official binary, put it on a 2GB digital ocean VPS as before. This time, it got to 67GB of (mostly small) files added, but was killed because it was out of memory. 

Things are a lot better in this version, but it seems like there is still a memory leak somewhere.

Happy to publish the script which I've been using if anyone wants to reproduce this, it's archiving open source data (python packages).
